81-question-mark-chrysalis/technical_specification.md

```diff
@@ -0,0 +1,230 @@
+```
+
+
+`WorkflowRoutes`
+
+- Purpose: describe named routes that bind workflows, input schema types and produced output schema types, and whether outputs should be published.
+
+data structure:
+```python
+class WorkflowRoute(BaseModel):
+   id: uuid
+   name: str | None
+   input_schema_id: uuid
+   output_schema_id: uuid
+   workflow_id: uuid
+```
+
+
+`AnnotatedEMPacks`
+
+Purpose: store incoming and derived annotated EM datapacks that are to be processed or published.
+
+data structure:
+```python
+class AnnotatedEMPack(BaseModel):
+   id: uuid
+   schema_id: uuid
+   original_id: uuid | None
+   datapack: DataPack
+   annotation: dict
+```
+
+#### Implementation Details of Collections
+
+The implementation of the EM transformation service will involve the following key components.
+
+1. `Schemas`
+   
+   Consists of Schema objects
+   * `id` is a unique identifier of the schema
+   * `name` is a human readable name of the schema
+   * `original` is a boolean flag indicating whether the schema is an original or derived schema
+   * `version` is the version of the schema (aka type of the schema), initially None if it is not an original schema
+   * `content` is the schema in schemapack format, initially None if it is not an original schema
+   * `publish` is a boolean flag indicating whether the derived datapacks conforming to this schema should be published
+   * `order` is the order in a topological ordering of the schemas in the transformation graph, initially None if it is not an original schema
+   
+
+   This collection is populated manually with the defined ghga schemas.
+
+   These are original schemas (`original: true`) stored in the `Schemas` and serve as the entry point for data that will be transformed into Universal GHGA Models. EMIM schemas provide a stable interface for data providers and may be more flexible or permissive than internal canonical models to accommodate varied input sources.
+
+2. `Workflows`
+
+   Consists of TransformationWorkflow objects
+   * `id` is a unique identifier for a workflow
+   * `name` is a human readable name for the workflow
+   * `description` is a more verbose description of the workflow
+   * `workflow` is the definition of the workflow in metldata format
+
+   This collection is populated manually with the defined ghga transformation workflows.
+
+
+1. `WorkflowRoutes`
+
+   Consists of WorkflowRoute object
+   * `id` is a unique identifier for the workflow route
+   * `name` is a human readable name for the workflow route
+   * `input_schema_id` is the id of the input schema that the workflow route accepts
+   * `output_schema_id` is the id of the output schema that the workflow route produces 
+   * `workflow_id` is a workflow identifier that is to be applied on a data / schema  
+   
+
+   This collection is populated manually with the defined ghga workflow routes.  
+   An empty workflow_output_schema indicates the route requires schema derivation at startup.  
```

Comment:
In the new design, all Schemas should be defined, so the `output_schema_id` should exist (even if that Schema doesn't have a computed schemapack yet).